objects/src/notes/metadata.rs

```diff
@@ -132,11 +134,77 @@ impl Serializable for NoteMetadata {
 impl Deserializable for NoteMetadata {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
         let sender = AccountId::read_from(source)?;
-        let note_type = NoteType::read_from(source)?;
+        let (note_type, note_execution_hint) = unmerge_type_and_hint(source.read_u64()?)
+            .map_err(|err| DeserializationError::InvalidValue(err.to_string()))?;
         let tag = NoteTag::read_from(source)?;
         let aux = Felt::read_from(source)?;
 
-        Self::new(sender, note_type, tag, aux)
+        Self::new(sender, note_type, tag, note_execution_hint, aux)
             .map_err(|err| DeserializationError::InvalidValue(err.to_string()))
     }
 }
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Encodes `note_type` and `note_execution_hint` into a [u64] such that the resulting number has
+/// the following structure (from most significant bit to the least significant bit):
+///
+/// - Bits 39 to 36 (4 bits): NoteType
+/// - Bits 35 to 32 (4 bits): NoteExecutionHint tag
+/// - Bits 31 to 0 (32 bits): NoteExecutionHint payload
+pub fn merge_type_and_hint(note_type: NoteType, note_execution_hint: NoteExecutionHint) -> u64 {
+    let type_nibble = note_type as u64 & 0xF;
+    let (tag_nibble, payload_u32) = note_execution_hint.into_parts();
+
+    let tag_section = (tag_nibble as u64) & 0xF;
+    let payload_section = payload_u32 as u64;
+
+    (type_nibble << 36) | (tag_section << 32) | payload_section
+}
+
+pub fn unmerge_type_and_hint(value: u64) -> Result<(NoteType, NoteExecutionHint), NoteError> {
```

Comment:
I'm wondering if it is possible to keep these functions private to this module somehow.

Reply:
Made these private